sycl/include/CL/sycl/detail/device_impl.hpp

```diff
@@ -70,163 +166,19 @@ class device_impl {
-    return createSyclObjFromImpl<platform>(
-      std::make_shared<platform_impl_pi>(plt));
-  }
-
-  bool has_extension(const string_class &extension_name) const override {
-    string_class all_extension_names =
-        get_device_info<string_class, info::device::extensions>::_(m_device);
-    return (all_extension_names.find(extension_name) != std::string::npos);
-  }
-
-  vector_class<device>
-  create_sub_devices(const cl_device_partition_property *Properties,
-                     size_t SubDevicesCount) const;
-
-  vector_class<device>
-  create_sub_devices(size_t ComputeUnits) const override;
-
-  vector_class<device>
-  create_sub_devices(const vector_class<size_t> &Counts) const override;
-
-  vector_class<device>
-  create_sub_devices(info::partition_affinity_domain AffinityDomain) const override;
+  is_affinity_supported(info::partition_affinity_domain AffinityDomain) const;
 
 private:
-  RT::PiDevice m_device = 0;
-  RT::PiDeviceType m_type;
-  bool m_isRootDevice = false;
-}; // class device_impl_pi
-
-// TODO: 4.6.4 Partitioning into multiple SYCL devices
-// TODO: 4.6.4.2 Device information descriptors
-// TODO: Make code thread-safe
-class device_host : public device_impl {
-public:
-  device_host() = default;
-  cl_device_id get() const override {
-    throw invalid_object_error("This instance of device is a host instance");
-  }
-  RT::PiDevice &getHandleRef() override {
-    throw invalid_object_error("This instance of device is a host instance");
-  }
-  const RT::PiDevice &getHandleRef() const override {
-    throw invalid_object_error("This instance of device is a host instance");
-  }
-
-  bool is_host() const override { return true; }
-
-  bool is_cpu() const override { return false; }
-
-  bool is_gpu() const override { return false; }
-
-  bool is_accelerator() const override { return false; }
-
-  platform get_platform() const override { return platform(); }
-
-  bool has_extension(const string_class &extension_name) const override {
-    // TODO: implement extension management;
-    return false;
-  }
-
-  vector_class<device> create_sub_devices(size_t nbSubDev) const override {
-    // TODO: implement host device partitioning
-    throw runtime_error(
-        "Partitioning to subdevices of the host device is not implemented yet");
-  }
-
-  vector_class<device>
-  create_sub_devices(const vector_class<size_t> &counts) const override {
-    // TODO: implement host device partitioning
-    throw runtime_error(
-        "Partitioning to subdevices of the host device is not implemented yet");
-  }
-
-  vector_class<device>
-  create_sub_devices(info::partition_affinity_domain affinityDomain) const override {
-    // TODO: implement host device partitioning
-    throw runtime_error(
-        "Partitioning to subdevices of the host device is not implemented yet");
-  }
-}; // class device_host
+  RT::PiDevice MDevice = 0;
+  RT::PiDeviceType MType;
+  bool MIsRootDevice = false;
+  bool MIsHostDevice;
```

Comment:
Can't we re-use `MType` field for storing this information?

Reply:
@AlexeySachkov PI doesn't know anything at all about host device:
https://github.com/intel/llvm/blob/84992a57cc85853f6cc6ff0036e7a260ba887bef/sycl/include/CL/sycl/detail/pi.h#L75-L79

And I think it's better to keep things going this way to avoid confusion.